tensorflow_probability/python/experimental/mcmc/covariance_reducer.py

```diff
@@ -0,0 +1,185 @@
+# Copyright 2020 The TensorFlow Probability Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ============================================================================
+"""CovarianceReducer for pre-packaged Streaming Covariance."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+# Dependency imports
+import tensorflow.compat.v2 as tf
+
+from tensorflow_probability.python.experimental.mcmc import reducer as reducer_base
+from tensorflow_probability.python.experimental.stats import sample_stats
+from tensorflow_probability.python.mcmc.internal import util as mcmc_util
+
+
+__all__ = [
+    'CovarianceReducer',
+    'VarianceReducer'
+]
+
+
+class CovarianceReducer(reducer_base.Reducer):
+  """`Reducer` that computes a running covariance.
+
+  As with all reducers, CovarianceReducer does not hold state information;
+  rather, it stores supplied metadata. Intermediate calculations are held in
+  a state object, which is returned via `initialize` and `one_step` method
+  calls.
+  """
+
+  def __init__(self, shape, event_ndims=None, dtype=tf.float32, name=None):
```

Comment:
Can we infer the shape(s) and dtype(s) from the first state that comes in to `initialize`, instead of requiring them here?

This would be a way the Reducer packaging would make this computation a little easier to use.

Also, can we support computing variances over structured latent spaces (like the Cox process model that has several separate Tensors in its latents)?  As a first simplification, let's assume that we're not trying to compute any cross-products across Tensors.  Then we should be able to get away with packing this Reducer with as many `RunningCovariance` instances as there are Tensors in the latent state (which becomes available to `initialize`), and letting them all operate independently.

Does this make sense?  Want to talk about it?